lib/wasi/src/runtime/module_cache/filesystem.rs

```diff
@@ -0,0 +1,231 @@
+use std::{
+    fs::File,
+    path::{Path, PathBuf},
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::{CacheError, Key, ModuleCache};
+
+/// A cache that saves modules to a folder on the host filesystem using
+/// [`Module::serialize()`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FileSystemCache {
+    cache_dir: PathBuf,
+}
+
+impl FileSystemCache {
+    pub fn new(cache_dir: impl Into<PathBuf>) -> Self {
+        FileSystemCache {
+            cache_dir: cache_dir.into(),
+        }
+    }
+
+    pub fn cache_dir(&self) -> &Path {
+        &self.cache_dir
+    }
+
+    fn path(&self, key: Key) -> PathBuf {
+        let artifact_version = wasmer_types::MetadataHeader::CURRENT_VERSION;
+        let dir = format!("artifact-v{artifact_version}");
+        self.cache_dir
+            .join(dir)
+            .join(key.to_string())
+            .with_extension("bin")
+    }
+}
+
+#[async_trait::async_trait]
+impl ModuleCache for FileSystemCache {
+    async fn load(&self, key: Key, engine: &Engine) -> Result<Module, CacheError> {
+        let path = self.path(key);
+
+        // FIXME: This will all block the thread at the moment. Ideally,
+        // deserializing and uncompressing would happen on a thread pool in the
+        // background.
+
+        let uncompressed = read_compressed(&path)?;
+
+        match Module::deserialize_checked(&engine, &uncompressed) {
+            Ok(m) => Ok(m),
+            Err(e) => {
+                tracing::debug!(
+                    %key,
+                    path=%path.display(),
+                    error=&e as &dyn std::error::Error,
+                    "Deleting the cache file because the artifact couldn't be deserialized",
+                );
+
+                if let Err(e) = std::fs::remove_file(&path) {
+                    tracing::warn!(
+                        %key,
+                        path=%path.display(),
+                        error=&e as &dyn std::error::Error,
+                        "Unable to remove the corrupted cache file",
+                    );
+                }
+
+                Err(CacheError::Deserialize(e))
+            }
+        }
+    }
+
+    async fn save(&self, key: Key, module: &Module) -> Result<(), CacheError> {
+        let path = self.path(key);
+
+        // FIXME: This will all block the thread at the moment. Ideally,
+        // serializing and compressing would happen on a thread pool in the
+        // background.
+
+        if let Some(parent) = path.parent() {
+            if let Err(e) = std::fs::create_dir_all(parent) {
+                tracing::warn!(
+                    dir=%parent.display(),
+                    error=&e as &dyn std::error::Error,
+                    "Unable to create the cache directory",
+                );
+            }
+        }
```

Comment:
I think it would be good to do the "setup" only when creating the `FilesystemCache` (that is, calling `::new(cache_dir)`.

Similarly as we currently do in the FilesystemCache
https://github.com/wasmerio/wasmer/blob/master/lib/cache/src/filesystem.rs#L43-L78

Reply:
I think it's good to always ensure the parent dir exists. 

Otherwise you could delete the cache dir and deploy stops workings. 
(I do this regularly when working on deploy)

The overhead should be small enough to not matter.